model/minsky.cc

```diff
@@ -713,6 +717,103 @@ namespace minsky
        });
     canvas.requestRedraw();
   }
+  
+  /*
+  For using GSL non-linear least square fitting routines
+  patterned on first Large Nonlinear Least Squares Example.
+  See online documentation at https://www.gnu.org/software/gsl/doc/html/nls.html
+  */
+      
+namespace
+{	
+	
+   struct model_params {
+	   double *t;
+	   double *x_init;
+	   shared_ptr<RKdata> d;
+   };      
+	
+   int residual_func(const gsl_vector * x, void *params,
+          gsl_vector * f)
+   {
+	  if (params==NULL) return GSL_EBADFUNC;
+	   	  
+	  struct model_params *par = ((struct model_params *)params); 
+	  
+      int n;
+      int err = GSL_SUCCESS;
+      n = ValueVector::stockVars.size();
+      vector<double> xx(n);
+      
+      for (size_t i = 0; i < n; i++)
+         xx[i] = gsl_vector_get(x,i); 
+       
+      // would like integration to continue from Minsky::step(). So, still a WORK IN PROGRESS
+      
+      //err=gsl_odeiv2_driver_apply(par->d->driver, par->t,numeric_limits<double>::max(),&xx[0]);
+	
+	  // define residual vector function to be optimized 
+      for (size_t i = 0; i < n; i++)
+          gsl_vector_set (f, i, xx[i] - par->x_init[i]);  
+            
+      return err;
+   }
+   
+      static void errHandlerOpt(const char* reason, const char* file, int line, int gsl_errno) {
+         throw error("gsl: %s:%d: %s",file,line,reason);
+   } 
+
+}      
+
+  int Minsky::LM_optimize(double x_out[], void * opt_pars)
+  {
+    	
+      gsl_set_error_handler(errHandlerOpt);	
+      const gsl_multifit_nlinear_type *T = gsl_multifit_nlinear_trust;
+      gsl_multifit_nlinear_workspace *w;
+      gsl_multifit_nlinear_fdf fdf;
+      gsl_multifit_nlinear_parameters fdf_params = gsl_multifit_nlinear_default_parameters();	
+      
+      struct model_params *params = ((struct model_params *)opt_pars);
+  	
+      const size_t n = stockVars.size();   
+      const size_t p = n; 
+      
+      gsl_vector *f = gsl_vector_alloc(n);
+      gsl_vector_view x = gsl_vector_view_array (x_out, p);    
+      int status, info;
+      
+      const double xtol = 1e-8;
+      const double gtol = 1e-8;
+      const double ftol = 0.0;
+      
+      /* define the function to be minimized */
+      fdf.f = residual_func;
+      fdf.df = NULL; //func_df;   /* set to NULL for finite-difference Jacobian */
+      fdf.fvv = NULL;     /* not using geodesic acceleration */
+      fdf.n = n;
+      fdf.p = p;
+      fdf.params = params; 
+      
+      /* allocate workspace with default parameters */
+      w = gsl_multifit_nlinear_alloc (T, &fdf_params, n, p);
```

Comment:
This also remains to be fixed in my implementation. I will do so when I rework the residual function so that it calls Minsky::step(), as you have suggested.